src/audio/clap/event_list.cpp

```diff
@@ -0,0 +1,154 @@
+
+static uint32_t size(const struct clap_input_events* in)
+{
+	auto* list = static_cast<const EventList*>(in->ctx);
+	return static_cast<uint32_t>(list->Size());
+}
+
+static const clap_event_header_t* get(const struct clap_input_events* in,
+                                      uint32_t index)
+{
+	auto* list = static_cast<const EventList*>(in->ctx);
+	return list->Get(index);
+}
+
+static bool try_push([[maybe_unused]] const struct clap_output_events* out,
+                     [[maybe_unused]] const clap_event_header_t* event)
+{
+	// Unimplemented (false means the event could not be pushed into queue)
+	return false;
+}
+
+EventList::EventList()
+{
+	constexpr auto NumBytes  = 8192;
+	constexpr auto NumEvents = 1024;
+
+	event_data.reserve(NumBytes);
+	event_offsets.reserve(NumEvents);
+
+	sysex_data.reserve(MaxMidiSysExBytes);
+
+	input_events  = {this, &size, &get};
+	output_events = {this, &try_push};
+}
+
+void EventList::Clear()
+{
+	event_data.clear();
+	event_offsets.clear();
+
+	sysex_data.clear();
+}
+
+void EventList::AddMidiEvent(const std::vector<uint8_t>& msg,
+                             const uint32_t sample_offset)
+{
+	assert(msg.size() == 2 || msg.size() == 3);
+
+	clap_event_midi ev = {};
+
+	ev.header.space_id = CLAP_CORE_EVENT_SPACE_ID;
+	ev.header.type     = CLAP_EVENT_MIDI;
+	ev.header.time     = sample_offset;
+	ev.header.flags    = 0;
+	ev.header.size     = sizeof(ev);
+	ev.port_index      = 0;
+	ev.data[0]         = msg[0];
+	ev.data[1]         = msg[1];
+
+	if (msg.size() == 3) {
+		ev.data[2] = msg[2];
+	} else {
+		ev.data[2] = 0;
+	}
+
+	const auto offset = event_data.size();
+	event_offsets.emplace_back(offset);
+
+	const auto ev_start = reinterpret_cast<uint8_t*>(&ev);
+	const auto ev_end   = ev_start + sizeof(ev);
+
+	event_data.insert(event_data.end(), ev_start, ev_end);
+}
+
+void EventList::AddMidiSysExEvent(const std::vector<uint8_t>& msg,
+                                  const uint32_t sample_offset)
+{
+	assert(msg.size() > 0);
+
+	clap_event_midi_sysex ev = {};
+
+	ev.header.space_id = CLAP_CORE_EVENT_SPACE_ID;
+	ev.header.type     = CLAP_EVENT_MIDI_SYSEX;
+	ev.header.time     = sample_offset;
+	ev.header.flags    = 0;
+	ev.header.size     = sizeof(ev);
+	ev.port_index      = 0;
+	ev.buffer          = sysex_data.data() + sysex_data.size();
+	ev.size            = msg.size();
+
+	sysex_data.insert(sysex_data.end(), msg.begin(), msg.end());
+
+	const auto offset = event_data.size();
+	event_offsets.emplace_back(offset);
+
+	const auto ev_start = reinterpret_cast<uint8_t*>(&ev);
+	const auto ev_end   = ev_start + sizeof(ev);
+
+	event_data.insert(event_data.end(), ev_start, ev_end);
+}
+
+uint32_t EventList::Size() const
+{
+	return event_offsets.size();
+}
+
+const clap_event_header_t* EventList::Get(const uint32_t index) const
+{
+	if (index < event_offsets.size()) {
+		return reinterpret_cast<const clap_event_header_t*>(
+		        &event_data[event_offsets[index]]);
+	}
+	return nullptr;
+}
```

Comment:
Not really an option, and it would complicate things anyway so I'm not even going there. And there are lots of CLAP event types for the generic use case. 

Do realise what we're really dealing with here are heterogeneous arrays. Every element has a short header that describes the item's length in bytes. The "get" function returns a pointer to the header, so the received can look at the header, get the length of the item, then read the appropriate number of bytes after the header.

It would be nice to have a native heterogeneous array concept in high level languages, but there isn't. The "workaround" is to use object inheritance, then store instance pointers having a fixed length in an array/vector. But that involves non-contiguous memory locations which can be very cache unfriendly. So this is a good practical solution and it's not too complicated.

I'm 100% going to use this solution in my UI work for storing variable-length draw commands in a contiguous array (it's important because there could be tens of thousands of draw commands per frame at 60+ frames per second).

Reply:
For the record, the event list code is a simplified version of the event list used in the official CLAP example host:
https://github.com/free-audio/clap-helpers/blob/59791394dc26637d9425c2745233a979602be2a7/include/clap/helpers/event-list.hh

IMO even that's is a bit overcomplicated with their "simulated" heap, so I stripped it down to the bare minimum.

Reply:
>  It would be nice to have a native heterogeneous array concept in high level languages but there isn't.

I mean, Rust enums let you do it quite easily.  The only downside is it uses a bit more space.  Every element becomes the size of the largest thing you want to store plus it needs an extra hidden field to tag what type it is.  It's a lot better than doing it with inheritance though.

```rust
struct Foo {
    a: u32,
    b: u32
}

struct Bar {
    a: f64,
    b: f64
}

enum FooOrBar {
    FooType(Foo),
    BarType(Bar),
}

fn main() {
    let mut vec = Vec::new();
    vec.push(FooOrBar::FooType(Foo{a: 42, b: 31}));
    vec.push(FooOrBar::BarType(Bar{a: 72.1, b: 55.128}));
}
```

I thought `std::variant` was much this same concept but if it complicates things too much don't worry about it.

Reply:
`std::any`! (ducks for cover...)

Reply:
> what we're really dealing with here are heterogeneous arrays. Every element has a short header that describes the item's length in bytes. The "get" function returns a pointer to the header, so the received can look at the header, get the length of the item, then read the appropriate number of bytes after the header.

> It would be nice to have a native heterogeneous array concept in high level languages, but there isn't.

Good news: C++17 does have `<any>` that let you put hetereogenous types into STL containers, so avoids all the book keeping like you mentioned with the header. The other cool thing is it avoids pointers and allocated blobs; so the elements are "fat" and present in the array.

```c++
std::vector<std::any> container;
container.push_back(std::string("Hello, World!"));
container.push_back(3.14159);
container.push_back(Point{10, 20});
container.push_back(Person{"Alice", 30});
```

Here's a full demo:

```c++
#include <any>
#include <cstdio>
#include <string>
#include <vector>

struct Point {
	int x = {};
	int y = {};
	constexpr void print() const
	{
		printf("{x=%d, y=%d}\n", x, y);
	}
};

struct Person {
	std::string name = {};
	int age          = 0;
	constexpr void print() const
	{
		printf("{name=\"%s\", age=%d}\n", name.c_str(), age);
	}
};

template <typename T>
constexpr bool print_any(const std::any& element)
{
	if (element.type() != typeid(T)) {
		return false;
	}

	if constexpr (std::is_same_v<T, std::string>) {
		printf("%s\n", std::any_cast<T>(element).c_str());
	} else if constexpr (std::is_same_v<T, double>) {
		printf("%f\n", std::any_cast<T>(element));
	} else if constexpr (std::is_same_v<T, Point> || std::is_same_v<T, Person>) {
		std::any_cast<T>(element).print();
	}

	return true;
}

template <typename... Types>
constexpr void print_if_type(const std::any& element)
{
	std::initializer_list<bool>{print_any<Types>(element)...};
}

int main()
{
	std::vector<std::any> container = {};

	container.push_back(std::string("Hello, World!"));
	container.push_back(3.14159);
	container.push_back(Point{10, 20});
	container.push_back(Person{"Alice", 30});

	for (const auto& element : container) {
		print_if_type<std::string, double, Point, Person>(element);
	}
	return 0;
}
```

There are simpler way to deal with elements than the initializer stuff, but at least for performance-intensive code, this lets each types be linked up to the processing code at compile-time.

Reply:
Thanks guys, but the code stays like this. I'm not complicating this further for no clear benefits. The current implementation works, doesn't use esoteric C++ features, and it simple to understand and debug.

Reply:
yup - zero recommendation here.  Only mentioned it so you're aware that heterogeneous arrays and vectors are part of the C++ language itself.

Reply:
> yup - zero recommendation here. Only mentioned it so you're aware that heterogeneous arrays and vectors are part of the C++ language itself.

Yeah, I realised that after writing my comment 😄